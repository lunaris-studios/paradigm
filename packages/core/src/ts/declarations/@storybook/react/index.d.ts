import * as Abstract from "@paradigmjs/abstract";
import * as Addons from "@storybook/addons";
import * as SR from "@storybook/react";

import "@storybook/react";

declare module "@storybook/react" {
	/**
	 * Metadata to configure the stories for an abstract component.
	 */
	declare type StorybookComponent =
		| React.ComponentType
		| Abstract.AbstractComponent
		| Abstract.AbstractPureComponent;
	export declare type Meta<
		TComponent extends StorybookComponent = any,
		TArgs = Addons.DefaultArgs
	> = Addons.BaseMeta<TComponent> &
		Addons.Annotations<Component.Spinner, React.ReactElement<unknown>>;
}
