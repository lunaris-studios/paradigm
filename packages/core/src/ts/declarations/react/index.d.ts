import * as SC from "styled-components";

declare module "react" {
	interface Attributes {
		css?: SC.CSSProp;
	}
}
