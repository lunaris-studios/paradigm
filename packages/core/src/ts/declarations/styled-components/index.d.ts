import * as Protocol from "@paradigmjs/protocol";

import "styled-components";

declare module "styled-components" {
	export declare interface IDefaultTheme extends Protocol.Theme {}
}
