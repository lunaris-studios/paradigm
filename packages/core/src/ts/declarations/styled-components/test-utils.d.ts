declare module "styled-components/test-utils" {
	import * as SC from "styled-components";

	export declare function find(
		wrapper: Object,
		styledComponent: SC.StyledComponent,
	): Element;

	export declare function find(
		element: Element,
		styledComponent: SC.StyledComponent,
	): Element;

	export declare function findAll(
		element: Element,
		styledComponent: SC.StyledComponent,
	): NodeList;
}
