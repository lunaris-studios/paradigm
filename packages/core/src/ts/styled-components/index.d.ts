import * as Protocol from "@paradigmjs/protocol";

import "styled-components";

declare module "styled-components" {
	export interface DefaultTheme extends Protocol.Theme {}
}
