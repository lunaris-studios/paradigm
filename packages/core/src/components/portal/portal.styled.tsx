import * as Protocol from "@paradigmjs/protocol";
import * as React from "react";
import * as SC from "styled-components";

// re-import `styled-components` development mode DOM classnames.
import styled, { css } from "styled-components";

/**
 * Table of Contents
 *
 * [Portal]
 */

/**
 * [Portal]
 * - [Portal.Container(WRAPPER)]
 */

export interface Portal {
	Container: SC.StyledComponent<
		"div",
		any,
		IPortalContainerAttrs,
		keyof IPortalContainerAttrs
	>;
}

export const Portal = {} as Portal;

/**
 * [Portal.Container]
 */

interface IPortalContainerProps
	extends SC.ThemeProps<SC.DefaultTheme>,
		React.HTMLProps<HTMLDivElement> {}

interface IPortalContainerAttrs extends IPortalContainerProps {}

Portal.Container = styled("div").attrs(
	(props: IPortalContainerProps): IPortalContainerAttrs => ({
		...props,
	}),
)<IPortalContainerProps>`
	${Protocol.Snippets.cover()}
`;
