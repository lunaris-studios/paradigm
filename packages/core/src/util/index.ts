export * from "./_can-use-dom";
export * from "./_compose";
export * from "./_function-utils";
export * from "./_is-key-board-click";
export * from "./_is-node-env";
export * from "./_js-utils";
export * from "./_pipe";
export * from "./_react-utils";
export * from "./_remove-non-html-props";
export * from "./_test-util";

import clamp from "lodash/clamp";

export { clamp };
export { default as isNull } from "lodash/isNull";
export { default as camelCase } from "lodash/camelCase";
