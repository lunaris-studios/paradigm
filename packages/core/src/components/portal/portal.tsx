import * as React from "react";
import * as ReactDOM from "react-dom";
import * as SC from "styled-components";

import * as Common from "~/common";
import * as Util from "~/util";

import * as Styled from "./portal.styled";

/** Detect if `React.createPortal()` API method does not exist. */
const cannotCreatePortal = !Util.isFunction(ReactDOM.createPortal);

export interface IPortalProps extends Common.IProps {
	/**
	 * Callback invoked when the children of this `Portal` have been added to the DOM.
	 */
	onChildrenMount?: () => void;

	/**
	 * The HTML element that children will be mounted to.
	 * @default document.body
	 */
	container: HTMLElement | null;
}

export interface IPortalState {
	hasMounted: boolean;
}

const defaultProps = Object.freeze<IPortalProps>({
	container: typeof document !== "undefined" ? document.body : null,
});

const defaultState = Object.freeze<IPortalState>({
	hasMounted: false,
});

/**
 * This component detaches its contents and re-attaches them to document.body.
 * Use it when you need to circumvent DOM z-stacking (for dialogs, popovers, etc.).
 * Any class names passed to this element will be propagated to the new container element on document.body.
 */
export class Portal extends React.Component<IPortalProps, IPortalState> {
	public static displayName = `${Common.DISPLAYNAME_PREFIX}.Portal`;

	static readonly defaultProps: IPortalProps = defaultProps;
	public readonly state: IPortalState = defaultState;

	public render() {
		const { children, container } = this.props;
		// Only render `children` once this component has mounted in a browser environment, so they are
		// immediately attached to the DOM tree and can do DOM things like measuring or `autoFocus`.
		// See long comment on componentDidMount in https://reactjs.org/docs/portals.html#event-bubbling-through-portals
		if (
			cannotCreatePortal ||
			typeof document === "undefined" ||
			!this.state.hasMounted ||
			!container
		) {
			return null;
		} else {
			return ReactDOM.createPortal(
				<Styled.Portal.Container>{children}</Styled.Portal.Container>,
				container,
			);
		}
	}

	public componentDidMount() {
		if (!this.props.container) {
			return;
		}

		this.setState({ hasMounted: true }, this.props.onChildrenMount);
		if (cannotCreatePortal) {
			this.unstableRenderNoPortal();
		}
	}

	public componentDidUpdate() {
		if (cannotCreatePortal) {
			this.unstableRenderNoPortal();
		}
	}

	private unstableRenderNoPortal() {
		const container = document.createElement("div");
		container.setAttribute(
			"style",
			"position: absolute; top: 0; right: 0; bottom: 0; left: 0;",
		);
		ReactDOM.unstable_renderSubtreeIntoContainer(
			/* parentComponent */ this,
			<div>{this.props.children}</div>,
			container,
		);
	}
}
