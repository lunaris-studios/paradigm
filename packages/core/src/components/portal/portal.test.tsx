import * as Protocol from "@paradigmjs/protocol";
import * as React from "react";

import * as Util from "~/util";

import * as Component from "./portal";
import * as Styled from "./portal.styled";

// @ts-ignore
import { find } from "styled-components/test-utils";

describe("Portal", () => {
	it("renders correctly", () => {
		const node = Util.wrapper(<Component.Portal />);
		const PortalContainer = find(node.baseElement, Styled.Portal.Container);
		expect(PortalContainer).toMatchSnapshot();
	});
});
