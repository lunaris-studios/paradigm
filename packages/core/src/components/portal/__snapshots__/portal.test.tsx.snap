// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Portal renders correctly 1`] = `
.c0 {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
}

<div
  class="c0"
/>
`;
